src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,392 @@
+        {
+            this.logger.LogTrace("()");
+
+            var batch = new List<BlockPair>();
+
+            Task<BlockPair> dequeueTask = null;
+            Task timerTask = null;
+
+            while (!this.nodeLifetime.ApplicationStopping.IsCancellationRequested)
+            {
+                // Start new dequeue task if not started already.
+                dequeueTask = dequeueTask ?? this.blocksQueue.DequeueAsync();
+
+                // Wait for one of the tasks: dequeue or timer (if available) to finish.
+                Task task = (timerTask == null) ? dequeueTask : await Task.WhenAny(dequeueTask, timerTask).ConfigureAwait(false);
+
+                bool saveBatch = false;
+
+                try
+                {
+                    await task.ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    // Happens when node is shutting down or Dispose() is called.
+                    // We want to save whatever is in the batch before exiting the loop.
+                    saveBatch = true;
+
+                    this.logger.LogDebug("Node is shutting down. Save batch.");
+                }
+
+                // Save batch if timer ran out or we've dequeued a new block and reached the consensus tip 
+                // or the max batch size is reached or the node is shutting down.  
+                if (dequeueTask.Status == TaskStatus.RanToCompletion)
+                {
+                    BlockPair item = dequeueTask.Result;
+
+                    // Set the dequeue task to null so it can be assigned on the next iteration.
+                    dequeueTask = null;
+
+                    batch.Add(item);
+
+                    this.currentBatchSizeBytes += item.Block.GetSerializedSize();
+
+                    saveBatch = saveBatch || (item.ChainedHeader == this.chain.Tip) || (this.currentBatchSizeBytes >= BatchThresholdSizeBytes);
+                }
+                else
+                {
+                    // Will be executed in case timer ran out or node is being shut down.
+                    saveBatch = true;
+                }
+
+                if (saveBatch)
+                {
+                    if (batch.Count != 0)
+                    {
+                        await this.SaveBatchAsync(batch).ConfigureAwait(false);
+
+                        batch.Clear();
+                        this.currentBatchSizeBytes = 0;
+                    }
+
+                    timerTask = null;
+                }
+                else
+                {
+                    // Start timer if it is not started already.
+                    timerTask = timerTask ?? Task.Delay(BatchMaxSaveIntervalSeconds * 1000, this.nodeLifetime.ApplicationStopping);
+                }
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>Checks if repository contains reorged blocks and deletes them; saves batch on top.</summary>
```

Comment:
Lets add to this comment the fact that the last item in the batch is considered to be on the main chain, and may trigger a reorg in case the store is on a different chain.

```
/// <summary>
///    Checks if repository contains reorged blocks and deletes them; saves batch on top.
///    The last item in the list is considered to be on the current main chain and will be used to determine if a reorg is triggered
///</summary>
```

Reply:
Fixed